eth/executionclient/execution_client.go

```diff
@@ -234,6 +227,90 @@ func (ec *ExecutionClient) fetchLogsInBatches(ctx context.Context, startBlock, e
 	return logCh, errCh
 }
 
+// subdivideLogFetch handles log fetching with automatic subdivision on rate limit errors.
+// It first attempts a direct fetch, and if that fails with a rate limit error, it subdivides
+// the block range and tries again with smaller chunks.
+func (ec *ExecutionClient) subdivideLogFetch(ctx context.Context, q ethereum.FilterQuery) ([]ethtypes.Log, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-ec.closed:
+		return nil, ErrClosed
+	default:
+	}
+
+	logs, err := ec.client.FilterLogs(ctx, q)
+	if err == nil {
+		return logs, nil
+	}
+
+	if !isRPCRateLimitError(err) || q.FromBlock == nil || q.ToBlock == nil {
+		return nil, err
+	}
+
+	fromBlock := q.FromBlock.Uint64()
+	toBlock := q.ToBlock.Uint64()
+
+	// check if we have at least 2 blocks
+	if fromBlock >= toBlock {
+		return nil, err
+	}
```

Comment:
fromBlock >= toBlock should be a new error :)

according to the comment we should also return err if less then 2 blocks?

Reply:
I added a new error. Also, the new comment clarifies the intention precisely